lib/os/queued_operation.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sys/queued_operation.h>
+#include <stdint.h>
+
+static inline int op_get_priority(const struct queued_operation *op)
+{
+	return (s8_t)(op->notify.flags >> QUEUED_OPERATION_PRIORITY_POS);
+}
+
+static inline int op_set_priority(struct queued_operation *op,
+				  int priority)
+{
+	s8_t prio = (s8_t)priority;
+	u32_t mask = (QUEUED_OPERATION_PRIORITY_MASK << QUEUED_OPERATION_PRIORITY_POS);
+
+	if (priority == QUEUED_OPERATION_PRIORITY_PREPEND) {
+		prio = INT8_MIN;
+	} else if (priority == QUEUED_OPERATION_PRIORITY_APPEND) {
+		prio = INT8_MAX;
+	} else if (prio != priority) {
+		return -EINVAL;
+	}
+
+	op->notify.flags = (op->notify.flags & ~mask)
+			   | (mask & (prio << QUEUED_OPERATION_PRIORITY_POS));
+
+	return 0;
+}
+
+static inline bool can_start(/*const*/ struct queued_operation_manager *mgr)
+{
+	/* OK if not busy, not finalizing, and an op is available. */
+	return (mgr->current == NULL)
+	       && !mgr->finalizing
+	       && !sys_slist_is_empty(&mgr->operations);
+}
+
+static void select_next_and_unlock(struct queued_operation_manager *mgr,
+				   k_spinlock_key_t key)
+{
+	struct queued_operation *op = NULL;
+
+	/* Track whether the manager is idle, so we only send
+	 * notification of entry to idle once.
+	 */
+	bool in_idle = (mgr->current == NULL);
+
+	do {
+		/* Can't select a new operation if one is active. */
+		if (mgr->current != NULL) {
+			break;
+		}
+
+		/* Can't select a new operation until the previous one
+		 * completes finalization, lest we pick something that
+		 * has too low a priority.
+		 */
+		if (mgr->finalizing) {
+			break;
+		}
+
+		sys_snode_t *node = sys_slist_get(&mgr->operations);
+
+		if (node) {
+			op = CONTAINER_OF(node, struct queued_operation, node);
+			mgr->current = op;
+		} else {
+			op = NULL;
+		}
+
+		k_spin_unlock(&mgr->lock, key);
+
+		/* Only notify the manager if there's an operation, or
+		 * if it is to transition to idle.
+		 */
+		if ((op != NULL) || !in_idle) {
+			mgr->vtable->process(mgr, op);
+		}
+		in_idle = (op == NULL);
+
+		key = k_spin_lock(&mgr->lock);
+	} while (op != NULL);
```

Comment:
If an operation is found (`op` is not null) and the call to `process()` finalizes the operation, then the service is available but has not been informed of this fact by re-invoking `process()` with a null pointer.  (There's a test that fails when the loop is not taken, showing this can happen.)

Reply:
This does not seem to be the right way of handling this case. The call to `process()` with a null pointer should be rather done in the context of `queued_operation_finalize()`. Currently this will not happen, because `in_idle` is `true` when `select_nex_and_unlock()` is called from `queued_operation_finalize()`. Consequently, `process()` is called with a null pointer in the case that you described, but not when `queued_operation_finalize()` is called asynchronously. For instance, the following test will fail:
```
static void test_async_processing(void)
{
	struct operation operation;

	reset_service();
	service.async = true;
	service.process_rv = 142;

	operation_init_spinwait(&operation);
	service_submit(&service, &operation, 0);
	service_finalize(&service, service.process_rv);
	zassert_false(service.active, "service not idled");
}
```

Reply:
OK, I see that.  Logic updated so the idle notification is sent in post-finalization.  Thanks.

Reply:
And now you can get rid of the loop here and apply what I suggested in https://github.com/zephyrproject-rtos/zephyr/pull/22853#discussion_r381904466. See https://github.com/anangl/zephyr/commit/4d1021deb18bb9fe6cc174671156e458dd0e50b2. All existing tests will pass.

Reply:
I just realized that both the current implementation and what I proposed above will not properly handle the following case:
- `select_next_and_unlock()` called for a finalized operation is preempted right before it is to call `process()` with a null pointer, and the preempting code calls `queued_operation_submit`

This will result in `process()` being called to start the new operation and right after that this preempted call to `process()` (with a null pointer) will be done.

Reply:
Yes, that makes sense.  We have to release the lock before calling process; I'll look into this later today.

Reply:
I suspect the problem identified above is part of the reason I originally had this in a loop, though the original logic wasn't quite right either.

It's back in a loop: after process() has been invoked the manager needs to check for delayed finalizations and additions to the active operation queue while the manager was unlocked.  This is supported by a more robust four-state machine that can distinguish the notification phase (unlocked invocation of process()) from processing and finalizing.

I think this is right, but I haven't added a specific test case yet.  I won't get back to this for a while, but a review of the updated logic would be helpful.

Reply:
Yes, now I see the need for having a loop in `select_next_and_unlock()`. Hovewer, the logic used in this function can be made much simpler and thus easier to understand, see https://github.com/anangl/zephyr/commit/586b5674df42b8abd330add02694008616af89e4.
 Every iteration of the loop should result in a call to `process()`, either to announce a new operation or to signal that there is nothing more to do. Otherwise, if nothing new was added to the queue in the notification phase and the currently processed operation hasn't finalized in that phase, there is no need to execute the loop again. Or am I missing something?
Actually, the current definition of `notify`:
```
bool notify = (op != NULL) || (state != ST_IDLE);
```
makes it always `true`, as `op` can be `NULL` only when `state` is `ST_FINALIZING` (there's an assertion right above that checks this).

Reply:
Agreed: that was a residue of a time before `lock_and_check_reselect()` handled both the logic and the state setting.